core/invoker/src/main/scala/whisk/core/invoker/ActionLogDriver.scala

```diff
@@ -0,0 +1,120 @@
+            var hasErr = false
+            var truncated = false
+            var bytesSoFar = 0.B
+            val logLines = Buffer[LogLine]()
+            val lines = logMsgs.lines
+
+            // read whiles bytesSoFar <= limit when requireSentinel to try and grab sentinel if they exist to indicate completion
+            while (lines.hasNext && ((requireSentinel && bytesSoFar <= limit) || bytesSoFar < limit)) {
+                // if line does not parse, there's an error in the container log driver
+                Try(lines.next().parseJson.convertTo[LogLine]) match {
+                    case Success(t) =>
+                        // if sentinels are expected, do not account for their size, otherwise, all bytes are accounted for
+                        if (requireSentinel && t.log.trim != LOG_ACTIVATION_SENTINEL || !requireSentinel) {
+                            // ignore empty log lines
+                            if (t.log.nonEmpty) {
+                                bytesSoFar += t.log.sizeInBytes
+                                if (bytesSoFar <= limit) {
+                                    logLines.append(t)
+                                } else {
+                                    // chop off the right most bytes that overflow
+                                    val chopped = t.dropRight(bytesSoFar - limit)
+                                    if (chopped.log.length > 0) {
+                                        logLines.append(chopped)
+                                    }
+                                    truncated = true
+                                }
+                            }
+                        } else if (requireSentinel) {
+                            // there may be more than one sentinel in stdout/stderr (as logs may leak across activations
+                            // if for example there log limit was exceeded in one activation and the container was reused
+                            // to run the same action again (this is considered a feature - otherwise, must drain the logs
+                            // or destroy the container as if it errored)
+                            if (t.stream == "stdout") {
+                                hasOut = true
+                            } else if (t.stream == "stderr") {
+                                hasErr = true
+                            }
+                        }
+
+                    case Failure(t) =>
+                        // Drop lines that did not parse to JSON objects.
+                        // However, should not happen since we are using the json log driver.
+                        logging.error(this, s"log line skipped/did not parse: $t")
+                }
+            }
+
+            ((hasOut && hasErr) || !requireSentinel, truncated || lines.hasNext, logLines.toVector)
+        } else {
+            logging.warn(this, s"log message is empty")
+            (!requireSentinel, false, Vector())
```

Comment:
Is this case only needed for logging? The above "return" line should evaluate to the same response, so we could add logging above it if lines was empty to flatten the branch hierarchy a bit.